doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,167 @@
+here, we're rarely going to list files, only write and read. If we do need to list files, it'll be for some cleanup
+operation and should not be on any request path.
+
+A big benefit of content-addressed storage here is that we get deduping almost for free -- we may need to zero out
+content that changes on every page load (e.g. "This page requested at HH:MM"), but that's more of an optimization than a
+hard requrement. These savings would apply to both documents which remain unchanged between fetches and documents that
+are shared between records.
+
+The actual document collection service will take a file upload and respond with a simple payload denoting the URI that
+may be used to access the file.
+
+### Manifest Construction
+
+Manifests need to be signed, so we'll just use [RFC 7519](https://tools.ietf.org/html/rfc7519) JWT blobs, encoding the
+manifest's information in the JWT payload. As an example, the payload could resemble:
+
+```json
+{
+  "version": 1,
+  "uuid": "6e566397-6803-49ec-9280-62c00b368e3f",
+  "record_type": "USA/FL/Bay/Court",
+  "case_id": "20000001",
+  "collected_at_ms": 1592618026000,
+  "fetcher_id": "github.com/Police-Data-Accessibility-Project/Scrapers:USA/FL/Bay/Court/scraper/Scraper.py@1db20ce1e63768363d532d46a641722986f63524",
+  "documents": [
+    {
+      "label": "some pdf we'll have to ocr",
+      "content_type": "application/pdf",
+      "digest": "...",
+      "uri": "https://..."
+    }, {
+      "label": "party information",
+      "content_type": "text/html",
+      "digest": "...",
+      "uri": "https://..."
+    }
+  ]
+}	
+```
+
+Determining the PDAP-controlled URIs of the documents will require uploading them first, and then constructing the
+manifest.
+
+### Garbage Collection
+Since this design requires upload documents before writing out a manifest, it's possible that we'll upload docs and then
+fail (or neglect) to upload a manifest. This results in orphaned documents, which are just wasted storage. 
+
+To mitigate this, a periodic (daily? weekly?) GC job will run in order to clean up orphaned documents that have existed
+for a while (>1 day should be sufficient) without a parent manifest. The set of orphans may be constructed by
+enumerating all documents in all manifests and subtracting those from the set of all stored documents.
+
+## Caveats
+This is somewhat complex and may be tricky to implement. A way to mitigate this is to invest in easy-to-use canonical
+implementations in common languages with an API that makes misuse hard. For example, a Python API could resemble:
+
+```python
+with Record(record_type="USA/FL/Bay/Court", case_id="1337") as r:
+  r.AddHtml("label1", html_file1)
+  r.AddHtml("label2", html_file2)
+  r.AddPdf("label3", pdf_file)
+  r.commit()  # upload docs, make and upload manifest
+```
+
+## Appendix
+### A half-baked proposal for JWT-based provenance verification
```

Comment:
There's an additional benefit here that's not discussed where records or keys found to be compromised can be easily cleaned up by revoking the key that signed the manifest

Reply:
Agreed. Expanded upon that a bit at the bottom.